File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/datagram/PriceSender.java
package ch.softappeal.yass.tutorial.datagram;

import ch.softappeal.yass.remote.*;
import ch.softappeal.yass.tutorial.contract.*;

import java.net.*;
import java.nio.channels.*;
import java.util.*;
import java.util.concurrent.*;

public final class PriceSender {

    public static final String GROUP_BID = "239.1.1.1";
    public static final String GROUP_ASK = "239.1.1.2";
    public static final int PORT = 8989;

    public static void main(final String... args) throws Exception {
        final DatagramChannel channel = DatagramChannel.open();
        final Client clientBid =
            DatagramTransport.client(Config.MESSAGE_SERIALIZER, channel, new InetSocketAddress(GROUP_BID, PORT));
        final Client clientAsk =
            DatagramTransport.client(Config.MESSAGE_SERIALIZER, channel, new InetSocketAddress(GROUP_ASK, PORT));
        final PriceListener priceListenerBid = clientBid.proxy(Config.INITIATOR.priceListener);
        final PriceListener priceListenerAsk = clientAsk.proxy(Config.INITIATOR.priceListener);
        for (int value = 1; true; value++) {
            System.out.println("sending BID and ASK: " + value);
            priceListenerBid.newPrices(Collections.singletonList(new Price(123, value, PriceKind.BID)));
            priceListenerAsk.newPrices(Collections.singletonList(new Price(123, value, PriceKind.ASK)));
            TimeUnit.MILLISECONDS.sleep(100);
        }
    }

}

File: kotlin/tutorial/main/ch/softappeal/yass/tutorial/datagram/PriceSender.kt
package ch.softappeal.yass.tutorial.datagram

import ch.softappeal.yass.tutorial.contract.Initiator
import ch.softappeal.yass.tutorial.contract.MessageSerializer
import ch.softappeal.yass.tutorial.contract.Price
import ch.softappeal.yass.tutorial.contract.PriceKind
import java.net.InetSocketAddress
import java.nio.channels.DatagramChannel
import java.util.concurrent.TimeUnit

const val GroupBid = "239.1.1.1"
const val GroupAsk = "239.1.1.2"
const val Port = 8989

fun main(args: Array<String>) {
    val channel = DatagramChannel.open()
    val clientBid = client(MessageSerializer, channel, InetSocketAddress(GroupBid, Port))
    val clientAsk = client(MessageSerializer, channel, InetSocketAddress(GroupAsk, Port))
    val priceListenerBid = clientBid.proxy(Initiator.priceListener)
    val priceListenerAsk = clientAsk.proxy(Initiator.priceListener)
    var value = 1
    while (true) {
        println("sending BID and ASK: $value")
        priceListenerBid.newPrices(listOf(Price(123, value, PriceKind.BID)))
        priceListenerAsk.newPrices(listOf(Price(123, value, PriceKind.ASK)))
        TimeUnit.MILLISECONDS.sleep(100)
        value++
    }
}
